src/core.js

```diff
@@ -12,11 +12,31 @@ jQuery.fn.init = function( selector ) {
 		migrateWarn( "jQuery( '#' ) is not a valid selector" );
 		args[ 0 ] = selector = [];
 
-	} else if ( rattrHash.test( selector ) ) {
+	// Can't use .test here because rattrHash is //g and in global scope
+	} else if ( rattrHash.exec( selector ) ) {
 
 		// The nonstandard and undocumented unquoted-hash was removed in jQuery 1.12.0
-		// Note that this doesn't actually fix the selector due to potential false positives
-		migrateWarn( "Attribute selectors with '#' must be quoted: '" + selector + "'" );
+		// First see if qS thinks it's a valid selector, if so avoid a false positive
+		try {
+			document.querySelector( selector );
+		} catch ( err1 ) {
+
+			// Didn't *look* valid to qSA, warn and try quoting what we think is the value
+			migrateWarn( "Attribute selectors with '#' must be quoted: `" + selector + "`" );
```

Comment:
I think we can omit wrapping backticks for messages like this that end with the variable data.